frontend/lib/resolution/split-init.cpp

```diff
@@ -19,196 +19,121 @@
+  void handleOutFormal(const FnCall* ast, const AstNode* actual,
+                       const QualifiedType& formalType,
+                       RV& rv) override;
+  void handleInFormal(const FnCall* ast, const AstNode* actual,
+                      const QualifiedType& formalType,
+                      RV& rv) override;
+  void handleInoutFormal(const FnCall* ast, const AstNode* actual,
+                         const QualifiedType& formalType,
+                         RV& rv) override;
+  void handleConditional(const Conditional* cond) override;
+  void handleTry(const Try* t) override;
+  void handleScope(const AstNode* ast) override;
 };
 
-FindSplitInits::FindSplitInits(Context* context)
-  : context(context)
-{
-}
-
-void FindSplitInits::handleMention(ID dstId) {
-  if (scopeStack.empty()) return;
-
+struct SplitInitVarStatus {
   bool foundReturnsThrows = false;
   bool foundInitedVar = false;
   bool foundMentionedVar = false;
   bool foundDeclaredVar = false;
+};
+
+SplitInitVarStatus FindSplitInits::findVarStatus(ID varId) {
+  SplitInitVarStatus status;
+
+  if (scopeStack.empty()) return status;
+
   for (ssize_t i = scopeStack.size() - 1; i >= 0; i--) {
-    SplitInitFrame* frame = scopeStack[i].get();
+    VarFrame* frame = scopeStack[i].get();
     if (frame->returns || frame->throws) {
-      foundReturnsThrows = true;
+      status.foundReturnsThrows = true;
     }
-    if (frame->initedVars.count(dstId) > 0) {
-      foundInitedVar = true;
+    if (frame->initedVars.count(varId) > 0) {
+      status.foundInitedVar = true;
     }
-    if (frame->mentionedVars.count(dstId) > 0) {
-      foundMentionedVar = true;
+    if (frame->mentionedVars.count(varId) > 0) {
+      status.foundMentionedVar = true;
     }
-    if (frame->declaredEligibleVars.count(dstId) > 0) {
-      foundDeclaredVar = true;
+    if (frame->declaredVars.count(varId) > 0) {
+      if (frame->eligibleVars.count(varId) > 0) {
+        status.foundDeclaredVar = true;
+      }
       break;
     }
   }
 
-  if (!foundDeclaredVar) return;
+  return status;
+}
 
-  // if it wasn't already initialized or mentioned, we mark it mentioned
-  // in this frame.
-  if (!foundReturnsThrows && !foundInitedVar && !foundMentionedVar) {
-    SplitInitFrame* frame = scopeStack.back().get();
-    frame->mentionedVars.insert(dstId);
+void FindSplitInits::addInit(VarFrame* frame,
+                             ID varId, QualifiedType rhsType) {
+  bool inserted = frame->addToInitedVars(varId);
+  if (inserted) {
+    // if it was inserted into the set, also record order and type
+    frame->initedVarsVec.emplace_back(varId, rhsType);
+  }
+}
+void FindSplitInits::addDeclaration(VarFrame* frame,
+                                    const VarLikeDecl* ast) {
+  bool inserted = frame->addToDeclaredVars(ast->id());
+  if (inserted) {
+    if (ast->initExpression() == nullptr) {
+      frame->eligibleVars.insert(ast->id());
+    }
   }
 }
```

Comment:
Implementing custom logic on `VarFrame`s in subclasses is a slight breeze in the direction of the "more type safe" approach (which specializes the VarFrames somehow per-use). However, all things considered, it's probably still better to continue with the simple approach.